modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.cpp

```diff
@@ -0,0 +1,694 @@
+
+    Size size = out.size();
+    if (out.isContinuous())
+    {
+        size.width *= size.height;
+        size.height = 1;
+    }
+
+    const bool isShiftRequired = shift != 0;
+
+    const std::size_t channelsCount = in.size();
+
+    if (isShiftRequired)
+    {
+        for (int i = 0; i < size.height; ++i)
+        {
+            auto rowPtr = out.ptr<OutT>(i);
+            for (std::size_t c = 0; c < channelsCount; ++c)
+            {
+                const auto first = in[c];
+                const auto last = first + size.width;
+                auto dOut = ChannelsIt(rowPtr, c, channelsCount);
+                std::transform(first, last, dOut, [shift](InT val) -> OutT { return val >> shift; });
+                in[c] += size.width;
+            }
+        }
+    }
+    else
+    {
+        for (int i = 0; i < size.height; ++i)
+        {
+            auto rowPtr = out.ptr<OutT>(i);
+            for (std::size_t c = 0; c < channelsCount; ++c)
+            {
+                const auto first = in[c];
+                const auto last = first + size.width;
+                auto dOut = ChannelsIt(rowPtr, c, channelsCount);
+                std::copy(first, last, dOut);
+                in[c] += size.width;
+            }
+        }
+    }
+}
+
+template<typename InT>
+void copyToMat(std::vector<const InT*>&& in, Mat& out, uint8_t shift)
+{
+    switch (out.depth())
+    {
+    case CV_8U:
+        copyToMatImpl<uint8_t>(std::move(in), out, shift);
+        break;
+    case CV_16U:
+        copyToMatImpl<uint16_t>(std::move(in), out, shift);
+        break;
+    default:
+        CV_Error(Error::StsNotImplemented, "only depth CV_8U and CV16_U are supported");
+    }
+}
+
+template<typename InT, typename OutT>
+void copyFromMatImpl(const Mat& in, std::vector<OutT*>&& out)
+{
+    using ChannelsIt = ChannelsIterator<ConstItTraits<InT>>;
+
+    Size size = in.size();
+    if (in.isContinuous())
+    {
+        size.width *= size.height;
+        size.height = 1;
+    }
+
+    const std::size_t outChannelsCount = out.size();
+
+    for (int i = 0; i < size.height; ++i)
+    {
+        const InT* row = in.ptr<InT>(i);
+        for (std::size_t c = 0; c < outChannelsCount; ++c)
+        {
+            auto first = ChannelsIt(row, c, outChannelsCount);
+            auto last = first + size.width;
+            out[c] = std::copy(first, last, out[c]);
+        }
+    }
+}
+
+template<typename OutT>
+void copyFromMat(const Mat& in, std::vector<OutT*>&& out)
+{
+    switch (in.depth())
+    {
+    case CV_8U:
+        copyFromMatImpl<uint8_t>(in, std::move(out));
+        break;
+    case CV_16U:
+        copyFromMatImpl<uint16_t>(in, std::move(out));
+        break;
+    default:
+        CV_Error(Error::StsNotImplemented, "only depth CV_8U and CV16_U are supported");
+    }
+}
+
+void errorLogCallback(const char* msg, void* /* userData */)
+{
+    CV_LOG_ERROR(nullptr, cv::format("OpenJPEG2000: %s", msg));
```

Comment:
> CV_LOG_ERROR(nullptr

Please use `NULL` for logger tags. It is designed as a part of identifier (`nullptr` would be broken in the future).